### PR TITLE
Implement scanner runs and querying them

### LINKED ISSUE #237: feat(scan-job): implement job tracking
**Task Description**

We need to track scan/data collection jobs and their scope definitions to remediate issues once they are resolved automatically. Therefore, job tracking must be implemented to store the job's status and definition of scope. 

We furthermore make the following assumptions: 

- A job is only complete when it reports that back
- A job of a specific scope does always collect the same "information" when the state of the information does not change
    -  therefore, items that do not appear can be handled after finishing the Job as 
- Overlapping scan scopes may occur 

Relevant Entities: 

- User 
- Component
- Component Instance
- Issue Match 
- Component Version 

**Acceptance Criteria:**

- [ ] When a new scan starts, it should be able to insert a new job by providing the following: 
    - a scope definition (plain string, decided by scanner)
    - (the start date is implicitly assumed to be the submission time)
- [ ] When a scan finishes, it should be able to update an existing job by providing the following: 
   -  ID of the Job
   -  state (success)
- [ ] When a scan fails, it should be able to update an existing job by providing the following: 
   -  ID of the Job
   -  state ( failure )
   -  context (free text/JSON)
- [ ] The scanner identity is derived from the JWT token (token-based authentication) and is attached to the Job. 
- [ ] It's not possible to update the Job of another identity 
- [ ] It's not possible to remove the Job of another identity from an entity 
- [ ] It is possible to add a job to an existing component that is already attached to another job
- [ ] All entities need to be able to get attached to a Job during creation and update
    - [ ] GraphQL Mutations for adding a job to each entity do exist
    - [ ] GraphQL Mutations for removing a job from each entity do exist
    - [ ] Respective App Layer functionality exists for each entity
    - [ ] Respective Data Access Layer functionality exists for each entity
    - [ ] Respective, composite indexed cross tables do exist
- [ ] All added App Layer  entries need to emit respective events using the event handler

**Expected Test:**

**E2E**
- Starting a Job (happy case)
- Finish a Job (happy case)
- Starting multiple jobs without finishing/reporting failure from the same identity should fail 
- Starting a job, finishing it, and starting then another job should work (happy case)
- Report Failure of a Job (Happy case)
- Finish a Job of another Identity should not work
- Report Failure of a Job of another Identity should not work
- Adding a job to an entity (for each entity) (happy case)
- Removing a job from an entity ( for each entity) (happy case) 
- Adding a non-existent job to an entity should fail (for each entity) 
- Removing a non-existent job to an entity should fail (for each entity) 
- Adding a non-existent job to an entity should fail (for each entity) 
- Removing a job that is not your identity should fail (for each entity) 

**App Layer**
- Happy Case Unit tests similar to E2E
- Error Cases similar to above, and ensure:
   - Check that we emit specific errors for wrong identity test cases
   - Check that we emit specific errors for not existing job test cases
   - Check that we emit specific errors for not existing entity test cases 
   - Check that we emit specific errors for double job running cases

**Database**
- Happy Case Unit Tests for all Inserts/Updates
- Edge Case Unit Tests for all Inserts/Updates

Dependencies
- #344

### PR DESCRIPTION
**Pull request by copyonwrite**
## Description
This PR implements tracking ScannerRuns, end to end

## What type of PR is this? (check all applicable)

- [x] 🍕 Feature

## Related Tickets & Documents

- Closes #237 and #238

## Added tests?

- [x] 👍 yes

## Added to documentation?

- [x] 🤝 Documentation pages updated

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: internal/database/interface.go [hunk 1]
```diff
@@ -127,8 +127,13 @@ type Database interface {
 	UpdateComponentVersion(*entity.ComponentVersion) error
 	DeleteComponentVersion(int64, int64) error
 
-	CreateScannerRun(*entity.ScannerRun) (*entity.ScannerRun, error)
+	CreateScannerRun(*entity.ScannerRun) (bool, error)
```
Comment (MR2011) [inline internal/database/interface.go:130]: All `Create` functions return the object. We should be consistent and return it for the ScannerRun too.

File: .devcontainer/Dockerfile [hunk 2]
```diff
@@ -1,7 +1,20 @@
...
+FROM golang
+
+ARG USERNAME=vscode
+ARG USER_UID=1000
+ARG USER_GID=$USER_UID
+
+RUN groupadd --gid $USER_GID $USERNAME \
+    && useradd --uid $USER_UID --gid $USER_GID -m $USERNAME -s /bin/bash \
+    && apt-get update \
+    && apt-get install -y sudo silversearcher-ag vim-nox autojump aspell aspell-en \
```
Comment (MR2011) [inline .devcontainer/Dockerfile:13]: Are these packages required or just personal preferences?

File: internal/app/scanner_run/scanner_run_handler.go [hunk 3]
```diff
@@ -25,15 +25,15 @@ type ScannerRunHandlerError struct {
 	msg string
 }
 
-func (srh *scannerRunHandler) CreateScannerRun(sr *entity.ScannerRun) (*entity.ScannerRun, error) {
+func (srh *scannerRunHandler) CreateScannerRun(sr *entity.ScannerRun) (bool, error) {
 	_, err := srh.database.CreateScannerRun(sr)
 
 	if err != nil {
-		return nil, &ScannerRunHandlerError{msg: "Error creating scanner run"}
+		return false, &ScannerRunHandlerError{msg: "Error creating scanner run"}
```
Comment (MR2011) [inline internal/app/scanner_run/scanner_run_handler.go:32]: After our error discussions, we should wrap the error

File: internal/app/scanner_run/scanner_run_handler.go [hunk 4]
```diff
@@ -43,10 +43,57 @@ func (srh *scannerRunHandler) CompleteScannerRun(uuid string) (bool, error) {
 		return false, &ScannerRunHandlerError{msg: "Error updating scanner run"}
```
Comment (MR2011) [inline internal/app/scanner_run/scanner_run_handler.go:43]: After our error discussions, we should wrap the error

File: internal/app/scanner_run/scanner_run_handler.go [hunk 5]
```diff
@@ -43,10 +43,57 @@ func (srh *scannerRunHandler) CompleteScannerRun(uuid string) (bool, error) {
...
-	srh.eventRegistry.PushEvent(&UpdateScannerRunEvent{})
+	srh.eventRegistry.PushEvent(&UpdateScannerRunEvent{successfulRun: true})
+	return true, nil
+}
+
+func (srh *scannerRunHandler) FailScannerRun(uuid string, message string) (bool, error) {
+	_, err := srh.database.FailScannerRun(uuid, message)
+
+	if err != nil {
+		return false, &ScannerRunHandlerError{msg: "Error updating scanner run"}
```
Comment (MR2011) [inline internal/app/scanner_run/scanner_run_handler.go:54]: After our error discussions, we should wrap the error

File: internal/database/interface.go [hunk 6]
```diff
@@ -127,8 +127,13 @@ type Database interface {
...
 	CompleteScannerRun(string) (bool, error)
+	FailScannerRun(string, string) (bool, error)
```
Comment (MR2011) [inline internal/database/interface.go:132]: Why don't we use a `UpdateScannerRun` function instead of `CompleteScannerRun` and `FailScannerRun`

File: internal/database/mariadb/scanner_run.go [hunk 7]
```diff
@@ -61,3 +61,176 @@ func (s *SqlDatabase) CompleteScannerRun(uuid string) (bool, error) {
 
 	return true, nil
 }
+
+func (s *SqlDatabase) FailScannerRun(uuid string, message string) (bool, error) {
+	updateScannerRunQuery := `UPDATE ScannerRun 
+					SET
```
Comment (MR2011) [inline internal/database/mariadb/scanner_run.go:67]: Indentation went wrong?

File: internal/database/mariadb/scanner_run.go [hunk 8]
```diff
@@ -61,3 +61,176 @@ func (s *SqlDatabase) CompleteScannerRun(uuid string) (bool, error) {
...
+	updateScannerRunQuery := `UPDATE ScannerRun 
+					SET 
+						scannerrun_is_completed = FALSE,
+						scannerrun_end_run = current_timestamp()
+					WHERE 
+						scannerrun_uuid = ? AND
+						scannerrun_is_completed = TRUE`
+
+	insertScannerRunErrorQuery := `INSERT INTO ScannerRunError 
+										(scannerrunerror_scannerrun_run_id, error)
```
Comment (MR2011) [inline internal/database/mariadb/scanner_run.go:75]: Indentation went wrong?

File: internal/database/mariadb/scanner_run.go [hunk 9]
```diff
@@ -61,3 +61,176 @@ func (s *SqlDatabase) CompleteScannerRun(uuid string) (bool, error) {
...
+	return &sr, nil
+}
+
+func (s *SqlDatabase) GetScannerRuns(filter *entity.ScannerRunFilter) ([]entity.ScannerRun, error) {
+	filter = s.ensureScannerRunFilter(filter)
+
+	baseQuery := `
+		SELECT * FROM ScannerRun
+    `
+	queryArgs := []any{}
```
Comment (MR2011) [inline internal/database/mariadb/scanner_run.go:117]: We have a `buildQueryParameters()` function that helps building this

File: internal/database/mariadb/scanner_run.go [hunk 10]
```diff
@@ -61,3 +61,176 @@ func (s *SqlDatabase) CompleteScannerRun(uuid string) (bool, error) {
...
+	if filter.After != nil {
+		baseQuery += " scannerrun_run_id > ?"
+		queryArgs = append(queryArgs, *filter.After)
+	}
+
+	baseQuery += " ORDER BY scannerrun_run_id"
+
+	queryArgs = append(queryArgs, *filter.First)
+	baseQuery += " LIMIT ?"
+	rows, err := s.db.Query(baseQuery, queryArgs...)
```
Comment (MR2011) [inline internal/database/mariadb/scanner_run.go:149]: You can use the `performListScan()` function to execute the query

File: internal/database/interface.go [hunk 11]
```diff
@@ -127,8 +127,13 @@ type Database interface {
...
+	ScannerRunByUUID(string) (*entity.ScannerRun, error)
+	GetScannerRuns(*entity.ScannerRunFilter) ([]entity.ScannerRun, error)
+	GetScannerRunTags() ([]string, error)
+	ScannerRunsTotalCount() (int, error)
```
Comment (MR2011) [inline internal/database/interface.go:136]: We should follow the naming convention here too and use `CountScannerRuns`

Comment (MR2011) [inline internal/database/interface.go:136]: CountScannerRuns should also accept a filter

File: internal/database/mariadb/scanner_run_test.go [hunk 12]
```diff
@@ -50,4 +51,330 @@ var _ = Describe("ScannerRun", Label("database", "ScannerRun"), func() {
...
+
+				Expect(err).To(BeNil())
+				Expect(newsr.Tag).To(Equal(sr.Tag))
+			})
+		})
+	})
+	When("Creating a new ScannerRun", Label("ByUUID"), func() {
+		Context("and the database is empty", func() {
+			It("should be marked as failed correctly", func() {
+				sr := &entity.ScannerRun{
```
Comment (MR2011) [inline internal/database/mariadb/scanner_run_test.go:77]: In the entity package we maintain functions that create Test entities with random data. I suggest adding ScannerRun there too

File: internal/database/mariadb/scanner_run_test.go [hunk 13]
```diff
@@ -50,4 +51,330 @@ var _ = Describe("ScannerRun", Label("database", "ScannerRun"), func() {
...
+
+				res, err := db.GetScannerRuns(nil)
+
+				Expect(err).To(BeNil())
+				Expect(len(res)).To(Equal(1))
+			})
+		})
+	})
+
+	When("Two ScannerRuns where created", Label("None"), func() {
```
Comment (MR2011) [inline internal/database/mariadb/scanner_run_test.go:127]: In general, should we add the scanner runs to the db seeder? @dorneanu @drochow

File: internal/database/mariadb/scanner_run_test.go [hunk 14]
```diff
@@ -50,4 +51,330 @@ var _ = Describe("ScannerRun", Label("database", "ScannerRun"), func() {
...
+
+				res, err := db.GetScannerRuns(nil)
+
+				Expect(err).To(BeNil())
+				Expect(len(res)).To(Equal(2))
+			})
+		})
+	})
+
+	When("Two ScannerRuns where created", Label("None"), func() {
```
Comment (MR2011) [inline internal/database/mariadb/scanner_run_test.go:160]: Multiple typos with where which should be were

File: internal/e2e/scanner_run_query_test.go [hunk 15]
```diff
@@ -66,16 +65,189 @@ var _ = Describe("Creating ScannerRun via API", Label("e2e", "ScannerRun"), func
...
+
+		s.NonBlockingStart()
+		client = graphql.NewClient(fmt.Sprintf("http://localhost:%s/query", cfg.Port))
+	})
+
+	AfterEach(func() {
+		s.BlockingStop()
 	})
+
+	When("the database is empty", func() {
```
Comment (MR2011) [inline internal/e2e/scanner_run_query_test.go:215]: We need a test with data in the DB where we also check if the correct data was returned

File: internal/database/interface.go [hunk 16]
```diff
@@ -127,8 +127,13 @@ type Database interface {
 	UpdateComponentVersion(*entity.ComponentVersion) error
 	DeleteComponentVersion(int64, int64) error
 
-	CreateScannerRun(*entity.ScannerRun) (*entity.ScannerRun, error)
+	CreateScannerRun(*entity.ScannerRun) (bool, error)
 	CompleteScannerRun(string) (bool, error)
+	FailScannerRun(string, string) (bool, error)
+	ScannerRunByUUID(string) (*entity.ScannerRun, error)
```
[the remark this answers is not in the document]
Comment (drochow) [reply internal/database/interface.go:133]: I agree we so far avoided adding special case getters  like get by xyz

File: internal/database/mariadb/init/schema.sql [hunk 17]
```diff
@@ -470,6 +470,15 @@ create table if not exists ScannerRunIssueTracker
     scannerrunissuetracker_scannerrun_run_id int unsigned not null,
     scannerrunissuetracker_issue_id  int unsigned not null,
 
-    constraint fk_run_id foreign key (scannerrunissuetracker_scannerrun_run_id) references ScannerRun (scannerrun_run_id) on update cascade,
-    constraint fk_issue_id foreign key (scannerrunissuetracker_issue_id) references Issue (issue_id) on update cascade
+    constraint fk_srit_sr_id foreign key (scannerrunissuetracker_scannerrun_run_id) references ScannerRun (scannerrun_run_id) on update cascade,
+    constraint fk_srit_i_id foreign key (scannerrunissuetracker_issue_id) references Issue (issue_id) on update cascade
```
Comment (drochow) [inline internal/database/mariadb/init/schema.sql:474]: Why is the "issue tracker" of a scanner run combined with an issue? The issue is actually some abstraction, which is the least important to track from the use-case perspective. 

We actually want to track issue matches. As Issue Matches are not created by any scanner but by Heureka itself the most important entities to track for the scanner runs are going to be: 

ComponentVersionIssue 

and 

ComponentInstance 

for all other entities, it a nice to have but beneficial to the use case from my perspective. 

What am I missing?

File: internal/database/mariadb/scanner_run_test.go (hunk 13 above)
Comment (drochow) [reply internal/database/mariadb/scanner_run_test.go:127]: Well, why not :)

File: internal/database/mariadb/scanner_run_test.go [hunk 18]
```diff
@@ -50,4 +51,330 @@ var _ = Describe("ScannerRun", Label("database", "ScannerRun"), func() {
...
+		Context("and the database is empty", func() {
```
Comment (drochow) [inline internal/database/mariadb/scanner_run_test.go:161]: It seems like we have here a repeating creation of 2 ScannerRuns with 2 different cases: 
once with 2 different Tag and once with 2 times the same Tag. Why not grouping the test cases accordingly and utilizing the beforeEach behavior to reduce the redundant creation logic?

**Review by drochow — CHANGES_REQUESTED** [review]
Agree with @MR2011 comments, added an additional comment which is unclear to me.

File: internal/database/mariadb/init/schema.sql (hunk 17 above)
Comment (MR2011) [reply internal/database/mariadb/init/schema.sql:474]: I think the idea was that if the Scanner doesn't report an Issue back, then we can assume that the corresponding IssueMatches can be closed. However, this does not address the case where only a subset of ComponentInstances are patched, because the Issue will be reported back by the scanner and therefore no IssueMatches will be closed

File: .devcontainer/Dockerfile (hunk 2 above)
Comment (copyonwrite) [reply .devcontainer/Dockerfile:13]: I switched the DevContainer from microsoft go container to the official Golang docker container (line 4), which is barebone and lacks a lot of tools for daily usage, thus some common tools and utilities were added.

File: internal/app/scanner_run/scanner_run_handler.go (hunk 3 above)
Comment (copyonwrite) [reply internal/app/scanner_run/scanner_run_handler.go:32]: Correct me, if I am wrong, but we did not effectively settle on a unified way to handle errors as a standard for the project. As soon as we agree on a unified standard for that, I'll happily follow the project standard.

File: internal/app/scanner_run/scanner_run_handler.go (hunk 4 above)
Comment (copyonwrite) [reply internal/app/scanner_run/scanner_run_handler.go:43]: see above

File: internal/app/scanner_run/scanner_run_handler.go (hunk 5 above)
Comment (copyonwrite) [reply internal/app/scanner_run/scanner_run_handler.go:54]: see above

File: internal/database/interface.go (hunk 1 above)
Comment (copyonwrite) [reply internal/database/interface.go:130]: Correct me, if I am wrong, but other objects are returned to the clients or used after creation. That is not true for the ScannerRun and the ScannerRun object is not meant to be exposed at all at this point, because all interaction should be indirect via the uuid of the ScannerRun. So, IMHO returning the object would unnecessarily expose internals at that point, which would increase coupling, leak internals and promotos unidiomatic usage. WONTFIX.

File: internal/database/interface.go (hunk 6 above)
Comment (copyonwrite) [reply internal/database/interface.go:132]: Because no one is supposed to update a scannerrun. The client creates a scanner run and then either the scanner run completes or fails.

File: internal/database/interface.go (hunk 11 above)
Comment (copyonwrite) [reply internal/database/interface.go:136]: Fixed.

File: internal/database/mariadb/init/schema.sql (hunk 17 above)
Comment (copyonwrite) [reply internal/database/mariadb/init/schema.sql:474]: This has been discussed already and further, this will be fixed if it is wrong, once the auto closing is implemented, thous it is of no real consequence at this moment.

File: internal/database/mariadb/scanner_run.go [hunk 19]
```diff
@@ -61,3 +61,176 @@ func (s *SqlDatabase) CompleteScannerRun(uuid string) (bool, error) {
...
+
+	for i := 0; filter.Tag != nil && i < len(filter.Tag); i++ {
+		baseQuery += " scannerrun_tag = ?"
+		queryArgs = append(queryArgs, filter.Tag[i])
+		if i < len(filter.Tag)-1 {
+			baseQuery += " OR"
+		}
+	}
+
+	if filter.Completed {
```
[the remark this answers is not in the document]
Comment (copyonwrite) [reply internal/database/mariadb/scanner_run.go:129]: I know.

File: internal/database/mariadb/scanner_run.go (hunk 10 above)
Comment (copyonwrite) [reply internal/database/mariadb/scanner_run.go:149]: I am using Golangs SQL API here because it seems to me that performListScan does not support transactions.

File: internal/database/mariadb/scanner_run_test.go (hunk 12 above)
Comment (copyonwrite) [reply internal/database/mariadb/scanner_run_test.go:77]: ok.

File: internal/database/mariadb/scanner_run_test.go (hunk 18 above)
Comment (copyonwrite) [reply internal/database/mariadb/scanner_run_test.go:161]: Because of the age old developer wisdom to have 3 examples before abstracting away something.

File: internal/database/mariadb/scanner_run_test.go (hunk 14 above)
Comment (copyonwrite) [reply internal/database/mariadb/scanner_run_test.go:160]: Fixed

File: internal/e2e/scanner_run_query_test.go (hunk 15 above)
Comment (copyonwrite) [reply internal/e2e/scanner_run_query_test.go:215]: There are already tests for the scanneruns in the db layer!?

File: internal/database/interface.go (hunk 1 above)
Comment (drochow) [reply internal/database/interface.go:130]: Well, we use a convention and currently do it everywhere, and the question is, is your argument strong enough to break the convention? To be discussed?

File: .devcontainer/Dockerfile (hunk 2 above)
Comment (drochow) [reply .devcontainer/Dockerfile:13]: Okay, I'll summarize this answer as yes, it's personal preference...

File: internal/database/mariadb/scanner_run.go (hunk 10 above)
Comment (drochow) [reply internal/database/mariadb/scanner_run.go:149]: Oh well, good idea to add transactions to the performListScan

File: internal/database/interface.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (drochow) [reply internal/database/interface.go:133]: Great! We agree to disagree!  👍

**Review by drochow — APPROVED** [review]
Its not my preference, but I'm Ok with addressing things later. 

For some things we should probably create backlog items though
